source/common/config/datasource.cc

```diff
@@ -100,29 +100,36 @@ DynamicData::~DynamicData() {
   }
 }
 
-absl::string_view DynamicData::data() const {
+const std::string& DynamicData::data() const {
   const auto thread_local_data = slot_->get();
   return thread_local_data.has_value() ? *thread_local_data->data_ : EMPTY_STRING;
 }
 
-absl::string_view DataSourceProvider::data() const {
+const std::string& DataSourceProvider::data() const {
   if (absl::holds_alternative<std::string>(data_)) {
     return absl::get<std::string>(data_);
   }
   return absl::get<DynamicData>(data_).data();
 }
 
-absl::StatusOr<DataSourceProvider> DataSourceProvider::create(const ProtoDataSource& source,
-                                                              Event::Dispatcher& main_dispatcher,
-                                                              ThreadLocal::SlotAllocator& tls,
-                                                              Api::Api& api, bool allow_empty,
-                                                              uint64_t max_size) {
+absl::StatusOr<DataSourceProviderPtr> DataSourceProvider::create(const ProtoDataSource& source,
+                                                                 Event::Dispatcher& main_dispatcher,
+                                                                 ThreadLocal::SlotAllocator& tls,
+                                                                 Api::Api& api, bool allow_empty,
+                                                                 uint64_t max_size) {
   auto initial_data_or_error = read(source, allow_empty, api, max_size);
   RETURN_IF_STATUS_NOT_OK(initial_data_or_error);
 
   if (!source.has_watched_directory() ||
       source.specifier_case() != envoy::config::core::v3::DataSource::kFilename) {
-    return DataSourceProvider(std::move(initial_data_or_error).value());
+    if (max_size > 0 && source.specifier_case() != envoy::config::core::v3::DataSource::kFilename &&
```

Comment:
Hah, we back to here.

Why the `max_size > 0` is necessary? Seems now the inline_string and inline_bytes have different length validation with the file?

Reply:
That is becuse uint64_t max_size = 0 has default value. We do not know whether it is set or default value is passed

Reply:
I see. Yeah, I agree with you about that the default value is a little weird.

But the read() also take a default 0, then let's keep the consistency.

If we do want to change it, we can do that in a independent PR to change both the read() and this method.